### PR TITLE
52 modelling temperature state in ates and constraining network temperature

### PR DESCRIPTION
**Pull request by FJanssen-TNO**
Start of the set up of ATES temperature variable and its constraints.
Still dummy equations are used and many other adaptation still required.
Main thoughts and ideas are in the code.

Temperature variables ates should only be created when multiple temperatures are available.

### CODE REVIEW

File: src/rtctools_heat_network/esdl/esdl_heat_model.py [hunk 1]
```diff
@@ -815,6 +817,17 @@ def convert_ates(self, asset: Asset) -> Tuple[Type[ATES], MODIFIERS]:
                 asset, "discountRate", default_value=0.0, min_value=0.0, max_value=100.0
             ),
             Q_nominal=self._get_connected_q_nominal(asset),
+            Q=dict(
+                min=-hfr_discharge_max / (4200 * 998 * dT) * asset.attributes["aggregationCount"],
+                max=hfr_discharge_max / (4200 * 998 * dT) * asset.attributes["aggregationCount"],
+                nominal=self._get_connected_q_nominal(asset) * asset.attributes["aggregationCount"],
```
Comment (jimrojerTNO) [inline src/rtctools_heat_network/esdl/esdl_heat_model.py:823]: We should get the cp and rho here

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 2]
```diff
@@ -1401,15 +1446,46 @@ def __pipe_heat_to_discharge_path_constraints(self, ensemble_member):
...
+            temperatures = self.temperature_regimes(carrier)
+
             for heat in [scaled_heat_in, scaled_heat_out]:
                 if self.heat_network_options()["neglect_pipe_heat_losses"]:
-                    constraints.append(
-                        (
-                            (heat - pipe_q * (cp * rho * temp)) / heat_nominal,
-                            0.0,
-                            0.0,
+                    if len(temperatures) == 0:
```
Comment (jimrojerTNO) [inline src/rtctools_heat_network/heat_physics_mixin.py:1454]: Do I see correctly that this is a bug fix? not specific to ATES

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 3]
```diff
@@ -1462,6 +1538,274 @@ def __pipe_heat_to_discharge_path_constraints(self, ensemble_member):
...
+                    )
+
+                    #linearisation of heatloss
+                    a, b = self.__get_linear_heatloss_vs_storedheat(heat_stored_max, ates_temperature,
+                                                                 temperature_ambient=soil_temperature)
+                    stored_heat_vec = ca.repmat(stored_heat, len(a))
+                    ates_heat_loss_vec = ca.repmat(heat_loss, len(a))
+                    ates_temperature_is_selected_vec = ca.repmat(ates_temperature_is_selected,
+                                                                 len(a))
+                    constraints.append(
```
Comment (jimrojerTNO) [inline src/rtctools_heat_network/heat_physics_mixin.py:1800]: Here we are adding energy loss constraints in the ATES where we:
- use inequality to approximate the quadratic increase in surface area of the hot bell and 
- use the discrete ATES temperature to model the heat transfer over the outer surface of the hot bell?

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 4]
```diff
@@ -1462,6 +1538,274 @@ def __pipe_heat_to_discharge_path_constraints(self, ensemble_member):
...
+                    # else:
+                    #     #if is selected, then specific temeprature loss constraint should be applicable, which will be a function of the stored heat
+                    #     # a(T_disc-Tamb) exp(-storedheat*b) where a and b are factors and the exponential function is linearised.
+                    a,b = self.__get_linear_Tloss_vs_storedheat(heat_stored_max, ates_temperature, temperature_ambient=soil_temperature)
+                    stored_heat_vec = ca.repmat(stored_heat, len(a))
+                    ates_dt_loss_vec = ca.repmat(ates_dt_loss, len(a))
+                    ates_temperature_is_selected_vec = ca.repmat(ates_temperature_is_selected, len(a))
+                    #TODO: improve nominal
+                    constraints.append(
+                        (( ates_dt_loss_vec - (a*stored_heat_vec/heat_stored_max + b) + big_m_dt*(1-ates_temperature_is_selected_vec))/ates_temperature_loss_nominal, 0.0 , np.inf)
```
Comment (jimrojerTNO) [inline src/rtctools_heat_network/heat_physics_mixin.py:1790]: I think this constraint might be wrong:
- Due to the inequality we might end up with positive losses in temperature, which are physically not possible

I think this cannot be done with inequality and should be a single linear equality constraint or piece-wise linear

File: src/rtctools_heat_network/pycml/component_library/heat/ates.py [hunk 5]
```diff
@@ -38,6 +39,10 @@ def __init__(self, name, **modifiers):
         self.rho = 988.0
         self.Heat_nominal = self.cp * self.rho * self.dT * self.Q_nominal
 
+        self.add_variable(Variable, "Temperature_ates", nominal=self.T_return)
+        self.add_variable(Variable, "Temperature_loss", min=0, max=1, nominal=self.T_supply/3600)
```
Comment (jimrojerTNO) [inline src/rtctools_heat_network/pycml/component_library/heat/ates.py:43]: I think the bounds are incorrect here, should be min=-1, max=0. I suppose

I think the nominal of T_supply/3600 is way too high, I would expect something in the order of a temperature change over a half year.

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 6]
```diff
@@ -1462,6 +1538,274 @@ def __pipe_heat_to_discharge_path_constraints(self, ensemble_member):
...
+            # dt change due to charging should be 0 when discharging. Because minimum bound is 0, ates_dt_charging can not be negative.
+            ates_dt_charging = self.state(f"{b}.Temperature_change_charging")
+            ates_dt_charging_nominal = self.variable_nominal(f"{b}.Temperature_change_charging")
+            big_m = bounds[f"{b}.Temperature_change_charging"][1]
+            # ensures no ates temperature change because of charging when discarging
+            constraints.append(((ates_dt_charging - is_buffer_charging * big_m)/ates_dt_charging_nominal, -np.inf, 0.0))
+
+            # #TODO temporary: still add relation to bound ates_dt_charging, should also be piecewise linear as a function of the stored heat, added heat and temperature difference
+            #maybe relation of tempeature_ates at first and last time step does not have to be the same for feasibility, maybe only the discrete variable.
+            constraints.append(((ates_dt_charging - .1 * heat_ates / heat_nominal - (
```
Comment (jimrojerTNO) [inline src/rtctools_heat_network/heat_physics_mixin.py:1631]: In my opinion the main reason to have the ATES temperature state is to have a better approximation of the operational cost during discharging in combination with the HP. I therefore think that at least for now we shouldn't invest in a stored_heat depending ates_dt_charging

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 7]
```diff
@@ -1462,6 +1538,274 @@ def __pipe_heat_to_discharge_path_constraints(self, ensemble_member):
...
+        The current stored heat is split in segments of equal size to linearise into n_lines
+        numbers of lines
+        heatloss/dt = a*stored_heat + b
+        """
+        #TODO: ensure heatloss and temperature loss linearisation rely on the same equations
+        heat_points = np.linspace(1.0, heat_max, n_lines+1) # cannot be 0
+
+        def algebraic_heatloss_ates(heat_points, heat_max, temperature_ates, temperature_ambient):
+            #TODO: function needs to be updated with realistic function, woudl normally non convex
+            heatloss = 1e-10*heat_points * (temperature_ates-temperature_ambient)*np.exp(-heat_points/heat_max)
```
Comment (jimrojerTNO) [inline src/rtctools_heat_network/heat_physics_mixin.py:1715]: I don't get this. I would expect a quadratic relation w.r.t. the stored_heat here, as the size of the bell increases the energy losses increase quadratically as the surface area of the bell increases quadratically.

**Review by jimrojerTNO — COMMENTED** [review]
This is the initial review, let's discuss before I go more in depth

File: src/rtctools_heat_network/heat_physics_mixin.py (hunk 7 above)
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/heat_physics_mixin.py:1715]: In reality it is also not a quadratic function:

in case of a sphere:
heat loss scales scales with the surface area (A ~r^2)
stored heat scales with the volume (V ~r^3)
thus heat loss ~V^(2/3) ~ stored heat ^(2/3)

however assuming it will be a cylinder and heat is only lossed on the sides, not to the above/below clay layers (assuming height constant):
heat loss scales scales with the surface area sides (A ~2pi r * h)
stored heat scales with the volume (V ~pi r^2 * h)
thus heat loss ~V^(1/2) ~ stored heat ^(1/2)

in case of a cylinder with heat losses on top and bottom dominant(assuming h constant):
heat loss scales scales with the surface area sides (A ~ 2pi r^2)
stored heat scales with the volume (V ~pi r^2 * h)
thus heat loss ~ V

in case of a cylinder with heat losses on all sides (assuming h constant):
heat loss scales scales with the surface area sides (A ~2pi r * h + 2pi r^2)
stored heat scales with the volume (V ~pi r^2 * h)
thus heat loss ~ (V + V^(1/2)) 

I'll check which function is most relevant, for now I'll keep an efficiency that changes with the temperature and will be multiplied with the stored heat

File: src/rtctools_heat_network/esdl/esdl_heat_model.py (hunk 1 above)
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/esdl/esdl_heat_model.py]: implemented

File: src/rtctools_heat_network/heat_physics_mixin.py (hunk 2 above)
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/heat_physics_mixin.py:1459]: Indeed, not of high priority so I thought it could just take it along in this development

File: src/rtctools_heat_network/pycml/component_library/heat/ates.py (hunk 5 above)
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/pycml/component_library/heat/ates.py]: improved, as well as the bounds

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 8]
```diff
@@ -2577,6 +3276,60 @@ def __heat_pump_cop_constraints(self, ensemble_member):
...
+                            np.inf,
+                        )
+                    )
+                    constraints.append(
+                        (
+                            (temperature - temperature_var - big_m * (1.0 - ordering_disc_carr)),
+                            -np.inf,
+                            0.0,
+                        )
+                    )
```
Comment (Tklauw-TNO) [inline src/rtctools_heat_network/heat_physics_mixin.py:3332]: These are actually ordering constraints on the selected supply temperature. I guess these constraints are relevant even for networks which allow selection of the supply temperature but don't have any ATES? If so, they should be moved to their own separate function and be called for every supply carrier.

**Review by Tklauw-TNO — COMMENTED** [review]
After a careful review I came to the following conclusions. Wanted to share here to make sure I understand correctly :).

In the current implementation I see that there are two temperatures defined for the ATES the regular one and a discrete one. However, one of the first constraints introduced for the ATES forces these two temperatures to be equal. This leads me to believe it might be more efficient to use a single ATES temperature. 

It should be possible to have a set of discrete supply temperatures and a continuous ATES temperature. The losses that change the continuous ATES temperature could be estimated based on the used supply temperature for that timestep. Difficulty here of course would be that there is a trade-off between the efficiency of the heat pump and the heat and temp losses of the ATES when the supply temperature isn't exactly equal to the ATES temperature. This trade-off could cause either over- or underestimation of the actual temperature change. I don't know the details here, but I do wonder if it's possible to force the optimizer into a situation that it is always conservative. I guess you might've already tried this approach based on the comments I saw though?

If the approach didn't work and thus shouldn't be used, that means the ATES temperature should be forced into a discrete state by the optimizer. In this case, I'm not sure if there are many good ways to improve the efficiency of the optimizer. I guess it will probably have some difficulties finding proper operation of the ATES since the temperature change is related to the losses, which is related to the temperature, which is discrete. So the optimizer will have to find, for every timestep, the proper operation of the ATES such that it can assign it temperature losses such that its temperature stays within the set of discrete values. Note that this is already not going great, if you run the test model of ates_temperatures, you see that the optimizer sets the actual ATES discrete temperature for 2 timesteps to ~65.7, which isn't exactly one of the allowed discrete temperatures... I think this is a clue that the optimizer is having some difficulty here in exactly matching the temperature of the ATES to one of the allowed discrete values due to the continuous dynamics.

TLDR: difficult problem, let's discuss ;)

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 9]
```diff
@@ -1542,6 +1684,526 @@ def __pipe_heat_to_discharge_path_constraints(self, ensemble_member):
...
+          has to provide, thus using more electricity and requiring a larger heatpump capacity.
+          Which will thus result in an over-estimation of the electricity used, an overestimation
+          in the required heat pump capacity (in case of sizing) or even a lower limit on the heat
+          that the heat pump can produce from the ATES.
+          - Finally a lower temperature also results in a lower heat transport capacity of the
+          pipes and a smaller heat extraction at those time steps due to the limit on the volumetric
+           flow in the ATES. This will result in a longer time for the extraction of heat from the
+           ATES to empty it, which will again result in more heat losses.
+
+        During unloading of the ATES, the temperature of the pipe should be the same as the
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/heat_physics_mixin.py:1775]: updated

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 10]
```diff
@@ -1542,6 +1684,526 @@ def __pipe_heat_to_discharge_path_constraints(self, ensemble_member):
...
+        discretized temperature of the ATES. During loading the temperature of the carrier/pipe
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/heat_physics_mixin.py:1776]: updated

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 11]
```diff
@@ -1542,6 +1684,526 @@ def __pipe_heat_to_discharge_path_constraints(self, ensemble_member):
...
+        should larger or equal to the discretized temperature of the ates.
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/heat_physics_mixin.py:1777]: updated

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 12]
```diff
@@ -1542,6 +1684,526 @@ def __pipe_heat_to_discharge_path_constraints(self, ensemble_member):
...
+        """
+
+        constraints = []
+        parameters = self.parameters(ensemble_member)
+
+        for b, (
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/heat_physics_mixin.py:1784]: updated

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 13]
```diff
@@ -1542,6 +1684,526 @@ def __pipe_heat_to_discharge_path_constraints(self, ensemble_member):
...
+            flow_dir_var = self._pipe_to_flow_direct_map[hot_pipe]
+            is_buffer_charging = self.state(flow_dir_var) * _hot_pipe_orientation
+
+            sup_carrier = parameters[f"{b}.T_supply_id"]
+            supply_temperatures = self.temperature_regimes(sup_carrier)
+            ates_temperature = self.state(f"{b}.Temperature_ates")
+            ates_temperature_disc = self.state(f"{b}__temperature_ates_disc")
+
+            # discretized tempeature should alwyas be smaller or equal to ATES temperature
+            constraints.append((ates_temperature - ates_temperature_disc, 0.0, 0.0))
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/heat_physics_mixin.py]: Theses changes are part of a larger group of changes. Updated as discussed.

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 14]
```diff
@@ -1542,6 +1684,526 @@ def __pipe_heat_to_discharge_path_constraints(self, ensemble_member):
...
+            if len(supply_temperatures) == 0:
+                constraints.append((parameters[f"{b}.T_supply"] - ates_temperature_disc, 0.0, 0.0))
+            else:
+                """
+                This function adds constraints to ensure that only one temperature level is active
+                for the ates temperature. Furthermore, it sets the
+                temperature variable to the temperature associated with the temperature of the
+                integer variable. Potentially these lines could be a separate function.
+                """
+                sum = 0.0
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/heat_physics_mixin.py]: updated

File: src/rtctools_heat_network/pycml/component_library/heat/ates.py [hunk 15]
```diff
@@ -62,13 +77,19 @@ def __init__(self, name, **modifiers):
...
+        self.add_variable(
+            Variable,
+            "Stored_volume",
+            min=0.0,
+            nominal=self._typical_fill_time * self.Q_nominal,
+        )
         self.add_variable(Variable, "Q", nominal=self.Q_nominal)
         # For nicer constraint coefficient scaling, we shift a bit more error into
         # the state vector entry of `Heat_loss`. In other words, with a factor of
         # 10.0, we aim for a state vector entry of ~0.1 (instead of 1.0)
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/pycml/component_library/heat/ates.py]: updated

File: src/rtctools_heat_network/pycml/component_library/heat/ates.py [hunk 16]
```diff
@@ -62,13 +77,19 @@ def __init__(self, name, **modifiers):
...
-        self._heat_loss_error_to_state_factor = 10.0
+        self._heat_loss_error_to_state_factor = 1  # 10.0
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/pycml/component_library/heat/ates.py]: updated

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 17]
```diff
@@ -1542,6 +1684,527 @@ def __pipe_heat_to_discharge_path_constraints(self, ensemble_member):
...
+                            (
+                                (
+                                    ates_dt_loss_vec
+                                    - big_m * (1.0 * np.ones(len(a)) - is_buffer_charging_vec)
+                                )
+                                / ates_temperature_loss_nominal,
+                                -np.inf,
+                                0.0,
+                            )
+                        )
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/heat_physics_mixin.py:2183]: updated

File: src/rtctools_heat_network/heat_physics_mixin.py [hunk 18]
```diff
@@ -2577,6 +3276,60 @@ def __heat_pump_cop_constraints(self, ensemble_member):
...
+
+        parameters = self.parameters(ensemble_member)
+
+        for ates in self.heat_network_components.get("ates", []):
+
+            sup_carrier = parameters[f"{ates}.T_supply_id"]
+            supply_temperatures = self.temperature_regimes(sup_carrier)
+            if len(supply_temperatures) != 0:
+                big_m = 2.0 * max(supply_temperatures)
+                min_dt = abs(min(np.diff(supply_temperatures)))
```
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/heat_physics_mixin.py:3349]: It shouldn't happen, but updated to >1 to ensure it can not occur accidently.

File: src/rtctools_heat_network/heat_physics_mixin.py (hunk 8 above)
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/heat_physics_mixin.py:3388]: agreed marked as a TODO as it requires slightly larger effort.

**Comment by FJanssen-TNO** [comment]
Several updated added as discussed, to improve the functionality and robustness. The PR should now be ready for a merge.

File: src/rtctools_heat_network/heat_physics_mixin.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Tklauw-TNO) [reply src/rtctools_heat_network/heat_physics_mixin.py:3349]: In the diff commit I see now it still looks like it's >0, not >1

File: tests/models/ates_temperature/src/run_ates_temperature.py [hunk 19]
```diff
@@ -353,14 +354,11 @@ def read(self):
...
-    print("T_ates_55: ", results["ATES_cb47__temperature_disc_55.0"])
-    print("T_ates_60: ", results["ATES_cb47__temperature_disc_60.0"])
-    print("T_ates_65: ", results["ATES_cb47__temperature_disc_65.0"])
-    print("T_ates_70: ", results["ATES_cb47__temperature_disc_70.0"])
-    print(f"time: {time.time() - t0}")
+    for i in range(0, 21):
+        print(f"T_ates_{40+i*1.5} :", results[f"ATES_cb47__temperature_disc_{40+i*1.5}.0"])
+        print(f"T_carrier_{40 + i * 1.5} :", results[f"41770304791669983859190_{40 + i * 1.5}.0"])
+
     a = 1
```
Comment (Tklauw-TNO) [inline tests/models/ates_temperature/src/run_ates_temperature.py:364]: These were used for some checking. I guess they can be removed now?

File: tests/test_temperature_ates_hp.py
Comment (Tklauw-TNO) [inline tests/test_temperature_ates_hp.py:1]: I see a few lines commented out in the test. Are these left there for a specific reason?

File: src/rtctools_heat_network/heat_physics_mixin.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (FJanssen-TNO) [reply src/rtctools_heat_network/heat_physics_mixin.py:3349]: My mistake, updated, thanks for your sharpness!

File: tests/test_temperature_ates_hp.py
Comment (FJanssen-TNO) [reply tests/test_temperature_ates_hp.py:1]: for the current example this check is not always true, might want to update in the future with a better example for which these hold.

**Review by Tklauw-TNO — APPROVED** [review]
